### PR TITLE
Implement ADMIN_DESIGN to choose a design

### PR DESCRIPTION
**Pull request by mikehaertl**
Solves #1.

I've also updated the README and removed the `timwolla/` prefix for the image there (the image is official now).

### CODE REVIEW

File: 4.3/Dockerfile [hunk 1]
```diff
@@ -4,7 +4,8 @@ STOPSIGNAL SIGINT
 
 RUN	addgroup -S adminer \
 &&	adduser -S -G adminer adminer \
-&&	mkdir -p /var/www/html
+&&	mkdir -p /var/www/html \
+&&	chown -R adminer:adminer /var/www/html
```
Comment (TimWolla) [inline 4.3/Dockerfile]: What's the purpose of this change?

File: 4.3/Dockerfile [hunk 2]
```diff
@@ -21,10 +22,20 @@ COPY	index.php .
 
 ENV	ADMINER_VERSION 4.3.0
 ENV	ADMINER_DOWNLOAD_SHA256 3584e1b3c7f853118d66a92f9c0996c8db39b05ac2a2b2623ff1676ec09bc08e
+ENV	ADMINER_SRC_DOWNLOAD_SHA256 1e82e9d4766c94a022c1f8e3357e8997ad03b29d2aae171ca2a10490342aee1f
 
 RUN	curl -fsSL https://www.adminer.org/static/download/$ADMINER_VERSION/adminer-$ADMINER_VERSION-en.php -o adminer.php \
 &&	echo "$ADMINER_DOWNLOAD_SHA256  adminer.php" |sha256sum -c -
 
+RUN	curl -fsSL https://github.com/vrana/adminer/archive/v$ADMINER_VERSION.tar.gz -o source.tgz \
```
Comment (TimWolla) [inline 4.3/Dockerfile]: Please merge this into the previous `RUN` line. This saves a layer in the resulting docker image. We cannot leverage caching here anyways, as both command chains are dependent on `ADMINER_VERSION` anyway.

Comment (TimWolla) [inline 4.3/Dockerfile]: Style: Use `.tar.gz` instead of `.tar.gz` please.

File: 4.3/Dockerfile [hunk 3]
```diff
@@ -21,10 +22,20 @@ COPY	index.php .
...
+&&	echo "$ADMINER_SRC_DOWNLOAD_SHA256  source.tgz" |sha256sum -c - \
+&&	tar xf source.tgz \
```
Comment (TimWolla) [inline 4.3/Dockerfile]: Please provide the `z` flag explicitely.

Comment (TimWolla) [inline 4.3/Dockerfile]: Please use `-C` to specify a target folder and provide a list of files to only extract the required files.

File: 4.3/entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+set -e
+
+if [[ -n "$ADMINER_DESIGN" ]]; then
```
Comment (TimWolla) [inline 4.3/entrypoint.sh]: Use `[` instead of `[[`. IIRC the latter is non-POSIX.

File: 4.3/entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+    ln -s "designs/$ADMINER_DESIGN/adminer.css" .
+fi;
```
Comment (TimWolla) [inline 4.3/entrypoint.sh]: And while you're at it: Remove the semicolon here.

File: 4.3/entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+set -e
+
+if [[ -n "$ADMINER_DESIGN" ]]; then
+    ln -s "designs/$ADMINER_DESIGN/adminer.css" .
```
Comment (TimWolla) [inline 4.3/entrypoint.sh]: To think about it, not sure on it myself: You `set -e`, but this is a non-critical line that may fail on read only file systems. Consider adding a `|| true`.

File: 4.3/Dockerfile (hunk 1 above)
Comment (mikehaertl) [inline 4.3/Dockerfile]: The `entrypoint.sh` runs as `admin`. Without this, the user has no permission to create a symlink.

File: 4.3/Dockerfile (hunk 2 above)
Comment (mikehaertl) [inline 4.3/Dockerfile]: Sure, no problem. I usually minimize the number of layers, too. Just thought, you prefer a more clear separation. Will update it.

File: 4.3/Dockerfile (hunk 3 above)
Comment (mikehaertl) [inline 4.3/Dockerfile]: Hmm, but why? We remove the complete folder anyway. I'd say we don't need any unneccessary options to keep it shorter and more clear.

File: 4.3/entrypoint.sh (hunk 6 above)
Comment (mikehaertl) [inline 4.3/entrypoint.sh]: I'd say: with the permission change on top `ln` in the container will always work. If someone maps a drive into the container and still has `ADMINER_DESIGN` set, it should throw an error to notify the user about the problem.

File: 4.3/Dockerfile (hunk 3 above)
Comment (mikehaertl) [inline 4.3/Dockerfile]: Ok, now I see what you mean. We can save one additional command. Will change it.

**Comment by mikehaertl** [comment]
@TimWolla I've addressed all your suggestions and pushed an update.

**Comment by mikehaertl** [comment]
@TimWolla One more thing: [Most](https://github.com/docker-library/php/blob/master/7.1/fpm/alpine/Dockerfile) [official](https://github.com/docker-library/node/blob/master/0.10/Dockerfile) [Dockerfiles](https://github.com/docker-library/mysql/blob/master/8.0/Dockerfile) seem to  usually indent chained commands on the next line. Is this something you could consider? If so I could update the MR accordingly.

**Comment by TimWolla** [comment]
> @TimWolla One more thing: Most official Dockerfiles seem to usually indent chained commands on the next line. Is this something you could consider? If so I could update the MR accordingly.

No. I prefer the current Dockerfile style with the Tab after the `&&` instead of before.

File: 4.3/Dockerfile (hunk 1 above)
Comment (TimWolla) [inline 4.3/Dockerfile]: Okay, sure.

File: 4.3/Dockerfile [hunk 7]
```diff
@@ -21,9 +22,17 @@ COPY	index.php .
...
 ENV	ADMINER_DOWNLOAD_SHA256 3584e1b3c7f853118d66a92f9c0996c8db39b05ac2a2b2623ff1676ec09bc08e
+ENV	ADMINER_SRC_DOWNLOAD_SHA256 1e82e9d4766c94a022c1f8e3357e8997ad03b29d2aae171ca2a10490342aee1f
 
 RUN	curl -fsSL https://www.adminer.org/static/download/$ADMINER_VERSION/adminer-$ADMINER_VERSION-en.php -o adminer.php \
-&&	echo "$ADMINER_DOWNLOAD_SHA256  adminer.php" |sha256sum -c -
+&&	echo "$ADMINER_DOWNLOAD_SHA256  adminer.php" |sha256sum -c - \
+&&	curl -fsSL https://github.com/vrana/adminer/archive/v$ADMINER_VERSION.tar.gz -o source.tar.gz \
+&&	echo "$ADMINER_SRC_DOWNLOAD_SHA256  source.tar.gz" |sha256sum -c - \
+&&	tar xzf source.tar.gz -C . --strip-components=1 "adminer-$ADMINER_VERSION/designs/" \
+&&	rm -rf source.tar.gz
```
Comment (TimWolla) [inline 4.3/Dockerfile]: Remove the `-rf`, please.

File: 4.3/Dockerfile [hunk 8]
```diff
@@ -21,9 +22,17 @@ COPY	index.php .
...
 ENV	ADMINER_VERSION 4.3.0
 ENV	ADMINER_DOWNLOAD_SHA256 3584e1b3c7f853118d66a92f9c0996c8db39b05ac2a2b2623ff1676ec09bc08e
+ENV	ADMINER_SRC_DOWNLOAD_SHA256 1e82e9d4766c94a022c1f8e3357e8997ad03b29d2aae171ca2a10490342aee1f
 
 RUN	curl -fsSL https://www.adminer.org/static/download/$ADMINER_VERSION/adminer-$ADMINER_VERSION-en.php -o adminer.php \
-&&	echo "$ADMINER_DOWNLOAD_SHA256  adminer.php" |sha256sum -c -
+&&	echo "$ADMINER_DOWNLOAD_SHA256  adminer.php" |sha256sum -c - \
+&&	curl -fsSL https://github.com/vrana/adminer/archive/v$ADMINER_VERSION.tar.gz -o source.tar.gz \
+&&	echo "$ADMINER_SRC_DOWNLOAD_SHA256  source.tar.gz" |sha256sum -c - \
+&&	tar xzf source.tar.gz -C . --strip-components=1 "adminer-$ADMINER_VERSION/designs/" \
```
Comment (TimWolla) [inline 4.3/Dockerfile]: I'm fairly sure `-C .` is redundant, because it's the implicit default. Please check and remove if I'm correct.

File: README.md [hunk 9]
```diff
@@ -10,16 +10,27 @@ Adminer (formerly phpMinAdmin) is a full-featured database management tool writt
...
 Note: This exposes the FastCGI socket to the Internet. Make sure to add proper firewall rules to prevent a direct access.
+
+### Choosing a design
+
+The image contains all the designs that are available in the source package of adminer.
+You can find an overview here: https://github.com/vrana/adminer/tree/master/designs.
+
+To activate a design you can pass its name in `ADMINER_DESIGN`:
+
+```
```
Comment (TimWolla) [inline README.md]: Please use the four-space-indentation style for code blocks. This is a requirement for official images.

File: 4.3/fastcgi/entrypoint.sh [hunk 10]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+set -e
+
+if [ -n "$ADMINER_DESIGN" ]; then
+    ln -s "designs/$ADMINER_DESIGN/adminer.css" .
```
Comment (TimWolla) [inline 4.3/fastcgi/entrypoint.sh]: Use a hard tab for consistency with the Dockerfile, please.e

**Comment by mikehaertl** [comment]
Updated again.

**Comment by TimWolla** [comment]
While testing I noticed that this does not work in combination with the FastCGI version of this Dockerfile. I'll have a look at it myself, but if you have a good solution: I'll be happy to hear.

**Comment by mikehaertl** [comment]
@TimWolla I now see, that the base image defines its own `ENTRYPOINT`: https://github.com/docker-library/php/blob/master/7.0/fpm/alpine/Dockerfile#L142

So instead of `exec "@"` we should execute `docker-php-entrypoint`. Just not sure how the correct `sh` syntax is for passing the arguments.

**Comment by TimWolla** [comment]
@mikehaertl Already fixed with d59b0c4d6b47fae722d1c4881715c1889ecd7f98 and bf71e7bbec99b2c3ece1a7ef0650c831bb6ecbb8.
